ui-tests/tst/software/aws/toolkits/jetbrains/uitests/fixtures/JTreeFixture.kt

```diff
@@ -7,53 +7,185 @@ import com.intellij.remoterobot.RemoteRobot
+
+    fun isPathSelected(vararg path: String) = collectSelectedPaths().contains(path.toList())
+
+    fun getValueAtRow(row: Int) = collectRows()[row]
+
+    fun expand(vararg path: String): JTreeFixture {
+        runJs("""
+            const expandingPathNodes = [${path.joinToString(",") { "\"${it}\"" }}]
+            const ignoreRoot = component.isRootVisible() === false
+            const treePath = com.intellij.ui.tree.TreePathUtil.convertArrayToTreePath(expandingPathNodes)
+            const toStringConverter = function(obj) {return java.util.Objects.toString(obj)}
+            const visitor = new com.intellij.ui.tree.TreeVisitor.ByTreePath(ignoreRoot, treePath, toStringConverter);
+            
+            com.intellij.util.ui.tree.TreeUtil.promiseExpand(component, visitor).blockingGet(5000)
+        """)
+        return this
+    }
+
+    fun expandAll(): JTreeFixture {
+        step("Expand all") {
+            runJs("com.intellij.util.ui.tree.TreeUtil.promiseExpandAll(component).blockingGet(5000)")
         }
+        return this
+    }
+
+    fun expandAllExcept(vararg nodes: String): JTreeFixture {
+        runJs("""
+            var excludedNodes = [${nodes.joinToString(",") { "\"$it\"" }}]
+            function visit(treePath) {
+                var pathStr = treePath.toString()
+                if (excludedNodes.some(function(node) { return pathStr.indexOf(node) !== -1 })) {
+                    return com.intellij.ui.tree.TreeVisitor.Action.SKIP_CHILDREN
+                } else {
+                    return com.intellij.ui.tree.TreeVisitor.Action.CONTINUE
+                }
+            }
+            const visitor = { visit: visit }
+            
+            com.intellij.util.ui.tree.TreeUtil.promiseExpand(component, new com.intellij.ui.tree.TreeVisitor(visitor)).blockingGet(5000)
+        """)
+        return this
+    }
+
+    fun collapsePath(vararg path: String, fullMatch: Boolean = true) = findExpandedPath(*path, fullMatch = fullMatch)?.let {
+        runJs("""
+            JTreeFixture(robot, component).collapseRow(${it.row})
+        """)
+    }
+
+    private fun findExpandedPath(vararg path: String, fullMatch: Boolean): TreePathToRow? = collectExpandedPaths().singleOrNull { expandedPath ->
+        expandedPath.path.size == path.size && expandedPath.path.containsAllNodes(*path, fullMatch = fullMatch) ||
+            expandedPath.path.size - 1 == path.size && expandedPath.path.drop(1).containsAllNodes(*path, fullMatch = fullMatch)
+    }
+
+    private fun findExpandedRowWithText(text: String, fullMatch: Boolean = true): TreePathToRow? = collectExpandedPaths().singleOrNull { expandedPath ->
+        if (fullMatch) expandedPath.path.last() == text else expandedPath.path.last().contains(text, true)
+    }
+
+    private fun List<String>.containsAllNodes(vararg treePath: String, fullMatch: Boolean): Boolean = zip(treePath).all {
```

Comment:
lots of things have changed here, but what is added to this newer version on top of what is forked?

Reply:
I'll add comments, its the JS blobs

heres the patch: https://github.com/JetBrains/intellij-ui-test-robot/issues/38#issuecomment-840973361